workflow/controller/controller.go

```diff
@@ -915,7 +915,17 @@ func (wfc *WorkflowController) tweakListRequestListOptions(options *metav1.ListO
 	options.LabelSelector = labelSelector.String()
 	// `ResourceVersion=0` does not honor the `limit` in API calls, which results in making significant List calls
 	// without `limit`. For details, see https://github.com/argoproj/argo-workflows/pull/11343
-	options.ResourceVersion = ""
+	// The reflector will record `lastSyncResourceVersion`. If `ResourceVersion != "0"`, we should use the value
+	// recorded by the reflector. see https://github.com/argoproj/argo-workflows/pull/13466
+	if options.ResourceVersion == "0" {
+		options.ResourceVersion = ""
+	}
+	// The reflector will set the Limit to `0` when `ResourceVersion != "" && ResourceVersion != "0"`, which will fail
+	// to limit the number of workflow returns. Timeouts and other errors may occur when there are a lots of workflows.
+	// see https://github.com/kubernetes/client-go/blob/ee1a5aaf793a9ace9c433f5fb26a19058ed5f37c/tools/cache/reflector.go#L286
```

Comment:
So in https://github.com/kubernetes/client-go/blob/ee1a5aaf793a9ace9c433f5fb26a19058ed5f37c/tools/cache/reflector.go#L289-L384 it says that no limit retrieves from watch cache and is intentionally done to avoid a thundering herd on etcd; so that sounds like it could have negative consequences?

Reply:
@agilgur5 Thank you for reading, let me share my opinion:
First of all, the purpose of this submission is to make the Limit take effect, in order to solve a series of problems caused by returning a large amount of workflow at once.
Next, the `ListAndWatch` method will only execute List once each time, and then start continuous Watch, so I think this operation will not cause too much load on etcd.
Summary: The ListWatch is not a frequent List request, Setting `Limit = xx` can ensure pagination return.

Reply:
Yea I get that the List portion is not as frequent (it will also happen during informer rebuilds), though part of the purpose of this PR is to reduce load on etcd, and this actually _causes_ a relist from etcd instead of the watch cache